google/cloud/forseti/services/inventory/storage.py

```diff
@@ -937,217 +906,50 @@ def write(self, resource):
 
         Args:
             resource (object): Resource object to store in db.
-
-        Raises:
-            Exception: If storage was opened readonly.
         """
-        if self.readonly:
-            raise Exception('Opened storage readonly')
-
-        resource_data = resource.data()
-        # Group members do not need to be checked if it already exists
-        # in Inventory, as a user can be members in multiple groups.
-        if resource_data.get('kind') != 'admin#directory#member':
-            previous_id = self._get_resource_id(resource)
-            if previous_id:
-                resource.set_inventory_key(previous_id)
-                self.update(resource)
+        # Use a lock to quickly check if this is a duplicate resource before
+        # updating the cache and proceeding.
+        with self._storage_lock:
+            if resource.get_full_resource_name() in self._wrote_resources:
```

Comment:
I see why the full name has to be diff here for policies, would keeping that data in the full name be the best approach or should we append the type at the end when checking instead?

Reply:
Having a single field that is unique for each row is useful. I'd rather not have to combine multiple fields to check uniqueness. This may change if we move closer to CAI's model in the future, but that can be addressed separately.